svm/tests/integration_test.rs

```diff
@@ -1,241 +1,538 @@
+    fn asserts(&self) -> Vec<TransactionBatchItemAsserts> {
+        self.transaction_batch
+            .iter()
+            .cloned()
+            .map(|item| item.asserts)
+            .collect()
+    }
 }
 
-#[test]
-fn svm_integration() {
+// one transaction in a batch plus check results for svm and asserts for tests
+#[derive(Clone, Debug)]
+pub struct TransactionBatchItem {
+    pub transaction: Transaction,
+    pub check_result: TransactionCheckResult,
+    pub asserts: TransactionBatchItemAsserts,
+}
+
+impl Default for TransactionBatchItem {
+    fn default() -> Self {
+        Self {
+            transaction: Transaction::default(),
+            check_result: Ok(CheckedTransactionDetails {
+                nonce: None,
+                lamports_per_signature: LAMPORTS_PER_SIGNATURE,
+            }),
+            asserts: TransactionBatchItemAsserts::succeeded(),
+        }
+    }
+}
+
+// asserts for a given transaction in a batch
+// we can automatically check whether it executed, whether it succeeded
+// log items we expect to see (exect match only), and rodata
+// rodata is a doubly-nested Option. None means no check is performed, Some(None) means we assert no rodata
+#[derive(Clone, Debug)]
+pub struct TransactionBatchItemAsserts {
+    pub executed: bool,
+    pub succeeded: bool,
+    pub logs: Vec<String>,
+    pub return_data: ReturnDataAssert,
+}
+
+impl TransactionBatchItemAsserts {
+    pub fn succeeded() -> Self {
+        Self {
+            executed: true,
+            succeeded: true,
+            logs: vec![],
+            return_data: ReturnDataAssert::Skip,
+        }
+    }
+
+    pub fn failed() -> Self {
+        Self {
+            executed: true,
+            succeeded: false,
+            logs: vec![],
+            return_data: ReturnDataAssert::Skip,
+        }
+    }
+
+    pub fn not_executed() -> Self {
+        Self {
+            executed: false,
+            succeeded: false,
+            logs: vec![],
+            return_data: ReturnDataAssert::Skip,
+        }
+    }
+
+    pub fn check_executed_transaction(&self, execution_details: &TransactionExecutionDetails) {
+        assert!(self.executed);
+        assert_eq!(self.succeeded, execution_details.status.is_ok());
+
+        if !self.logs.is_empty() {
+            let actual_logs = execution_details.log_messages.as_ref().unwrap();
+            for expected_log in &self.logs {
+                assert!(actual_logs.contains(expected_log));
+            }
+        }
+
+        match (&self.return_data, &execution_details.return_data) {
+            (ReturnDataAssert::Some(expected_ro_data), Some(actual_ro_data)) => {
+                assert_eq!(expected_ro_data, actual_ro_data)
+            }
+            (ReturnDataAssert::None, None) => {}
+            (ReturnDataAssert::Skip, _) => {}
+            (left, right) => panic!(
+                "assertion `left == right` failed\n  {:?}\n  {:?}",
+                left, right
+            ),
+        }
```

Comment:
I think you can simply do `assert_eq!(self.return_data, execution_details.return_data)`, provided that you derive `PartialEq` for `ReturnDataAssert`.

Reply:
cant because theyre different types, but there is a natural `From` impl!